### PR TITLE
A Very Productive Plane Ride

### PR DESCRIPTION
**Pull request by admoore0**
- Fixed dialog color references.
- Fixed bug with barricades and battering rams
- Removed `melee_unit_data`, etc. as subclasses.
  - Now all stats that can be displayed on a card are in the `CardData` base
  - As a bonus, this got rid of a bunch of `extra_data` members
  - Also added string overrides for more complicated display (see Lightning)
- Barricade unit tooltip now displays properly
- Trebuchet no longer has the weirdness with placement (open to reverting if you want)
- Units now move to the space a barricade was in after destroying it
- Added card pool resources for better weighting (weights open to rebalancing)
- *Now rounds that have multiple cards played actually get stored properly*

I have been playtesting this with family and it's clear the tutorial needs more work. They didn't know which one was their own castle, what the health was, the fact you can only place units in the staging area, etc. Out of scope for this merge/maybe this release, but something we should consider.

### CODE REVIEW

File: src/cards/actions/charge.gd [hunk 1]
```diff
@@ -38,17 +35,12 @@ func can_perform(grid_pos: Vector2i, is_enemy: bool) -> bool:
...
-
-	var extra_speed := int(data.extra_data[0])
-
 	game.play_sound(game.SoundEffect.CHARGE, not is_enemy)
 
 	var units = melee_units.get_children().filter(filter_func)
 	for unit in units as Array[MeleeUnit]:
-		unit.extra_stats["speed"] = extra_speed
+		# Do we want to be able to stack buffs?
+		unit.extra_stats["speed"] = data.movement
```
Comment (kdreese) [inline src/cards/actions/charge.gd]: I would say yes

File: src/cards/dual_card.gd [hunk 2]
```diff
@@ -71,28 +71,37 @@ func hide_all() -> void:
 
 
 func update_icons(data: CardData, grid: GridContainer) -> void:
-	if data is MeleeUnitData:
-		grid.get_node("HealthLabel").show()
-		grid.get_node("HealthLabel").text = str(data.health)
+	for node in grid.get_children():
+		node.hide()
+	if data.info_show_flags & CardData.SHOW_HEALTH_FLAG:
```
Comment (kdreese) [inline src/cards/dual_card.gd:76]: Should this be "and" instead of "&"?

File: src/cards/attack/cavalier_2.tres [hunk 3]
```diff
@@ -1,20 +1,31 @@
...
 [ext_resource type="Texture2D" uid="uid://cxb7tipkeyik5" path="res://assets/icons/horse.png" id="1_qlihe"]
-[ext_resource type="Script" path="res://src/cards/attack/melee_unit_data.gd" id="2_mq7k3"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="2_mq7k3"]
 
 [resource]
 script = ExtResource("2_mq7k3")
 attack_power = 3
 attack_recoil = 1
-health = 4
+health_unused = 0
```
Comment (kdreese) [inline src/cards/attack/cavalier_2.tres]: ??

File: src/cards/attack/cavalier_3.tres [hunk 4]
```diff
@@ -1,20 +1,31 @@
...
 [ext_resource type="Texture2D" uid="uid://cxb7tipkeyik5" path="res://assets/icons/horse.png" id="1_xqf13"]
-[ext_resource type="Script" path="res://src/cards/attack/melee_unit_data.gd" id="2_w7mb6"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="2_w7mb6"]
 
 [resource]
 script = ExtResource("2_w7mb6")
 attack_power = 4
 attack_recoil = 1
-health = 7
+health_unused = 0
```
Comment (kdreese) [inline src/cards/attack/cavalier_3.tres]: ??

File: src/cards/attack/soldier_1.tres [hunk 5]
```diff
@@ -1,19 +1,31 @@
...
 [ext_resource type="Texture2D" uid="uid://ce786en8ql3ui" path="res://assets/icons/sword.png" id="1_cmh0s"]
-[ext_resource type="Script" path="res://src/cards/attack/melee_unit_data.gd" id="2_brfab"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="2_brfab"]
 
 [resource]
 script = ExtResource("2_brfab")
 attack_power = 4
 attack_recoil = 1
-health = 4
+health_unused = 0
```
Comment (kdreese) [inline src/cards/attack/soldier_1.tres:10]: ??

File: src/cards/attack/soldier_2.tres [hunk 6]
```diff
@@ -1,19 +1,31 @@
...
 [ext_resource type="Texture2D" uid="uid://ce786en8ql3ui" path="res://assets/icons/sword.png" id="1_oolq4"]
-[ext_resource type="Script" path="res://src/cards/attack/melee_unit_data.gd" id="2_7uogh"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="2_7uogh"]
 
 [resource]
 script = ExtResource("2_7uogh")
 attack_power = 6
 attack_recoil = 1
-health = 7
+health_unused = 0
```
Comment (kdreese) [inline src/cards/attack/soldier_2.tres:10]: ??

File: src/cards/attack/soldier_3.tres [hunk 7]
```diff
@@ -1,19 +1,31 @@
...
 [ext_resource type="Texture2D" uid="uid://ce786en8ql3ui" path="res://assets/icons/sword.png" id="1_0200b"]
-[ext_resource type="Script" path="res://src/cards/attack/melee_unit_data.gd" id="2_jcl6a"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="2_jcl6a"]
 
 [resource]
 script = ExtResource("2_jcl6a")
 attack_power = 7
 attack_recoil = 1
-health = 10
+health_unused = 0
```
Comment (kdreese) [inline src/cards/attack/soldier_3.tres:10]: ??

File: src/cards/card_pool.gd [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+## Reset the state of the 'cards' variable. Call this before doing draws.
+func reset() -> void:
+	cards.clear()
+	for entry in entries:
+		for _idx in entry.quantity:
+			cards.append(entry.card)
+
+
+## Draw a new card and remove it from the pool.
+func draw_card() -> CardData:
```
Comment (kdreese) [inline src/cards/card_pool.gd]: Looks good but maybe reword to not confuse with drawing from the deck. Maybe add_card()?

File: src/cards/defense/archer_1.tres [hunk 9]
```diff
@@ -1,17 +1,29 @@
...
 
 [ext_resource type="Texture2D" uid="uid://b1v1l3sgmb7bd" path="res://assets/icons/bow.png" id="1_3bokg"]
-[ext_resource type="Script" path="res://src/cards/defense/ranged_unit_data.gd" id="1_onhje"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="1_onhje"]
 
 [resource]
 script = ExtResource("1_onhje")
 attack_range = 1
-attack_damage = 1
+attack_damage = 0
```
Comment (kdreese) [inline src/cards/defense/archer_1.tres]: ??

File: src/cards/defense/archer_2.tres [hunk 10]
```diff
@@ -1,17 +1,29 @@
...
 
 [ext_resource type="Texture2D" uid="uid://b1v1l3sgmb7bd" path="res://assets/icons/bow.png" id="1_gvjhh"]
-[ext_resource type="Script" path="res://src/cards/defense/ranged_unit_data.gd" id="2_5c4a7"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="2_5c4a7"]
 
 [resource]
 script = ExtResource("2_5c4a7")
 attack_range = 1
-attack_damage = 2
+attack_damage = 0
```
Comment (kdreese) [inline src/cards/defense/archer_2.tres]: ??

File: src/cards/defense/barricade_1.tres [hunk 11]
```diff
@@ -1,16 +1,28 @@
...
+[gd_resource type="Resource" script_class="CardData" load_steps=3 format=3 uid="uid://b5k015tk0h26j"]
 
-[ext_resource type="Script" path="res://src/cards/defense/barricade_unit_data.gd" id="1_5w8r7"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="1_5w8r7"]
 [ext_resource type="Texture2D" uid="uid://bcmughyc0xshu" path="res://assets/icons/barricade.png" id="1_yy665"]
 
 [resource]
 script = ExtResource("1_5w8r7")
-health = 8
+health_unused = 0
```
Comment (kdreese) [inline src/cards/defense/barricade_1.tres]: ??

File: src/cards/defense/barricade_2.tres [hunk 12]
```diff
@@ -1,16 +1,28 @@
...
+[gd_resource type="Resource" script_class="CardData" load_steps=3 format=3 uid="uid://y5kpu5x4y5xh"]
 
 [ext_resource type="Texture2D" uid="uid://bcmughyc0xshu" path="res://assets/icons/barricade.png" id="1_3up0x"]
-[ext_resource type="Script" path="res://src/cards/defense/barricade_unit_data.gd" id="2_cllte"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="2_cllte"]
 
 [resource]
 script = ExtResource("2_cllte")
-health = 12
+health_unused = 0
```
Comment (kdreese) [inline src/cards/defense/barricade_2.tres:8]: ??

File: src/cards/defense/barricade_3.tres [hunk 13]
```diff
@@ -1,16 +1,28 @@
...
+[gd_resource type="Resource" script_class="CardData" load_steps=3 format=3 uid="uid://4jigsl24yqat"]
 
 [ext_resource type="Texture2D" uid="uid://bcmughyc0xshu" path="res://assets/icons/barricade.png" id="1_43i26"]
-[ext_resource type="Script" path="res://src/cards/defense/barricade_unit_data.gd" id="2_fm1ay"]
+[ext_resource type="Script" path="res://src/cards/card_data.gd" id="2_fm1ay"]
 
 [resource]
 script = ExtResource("2_fm1ay")
-health = 20
+health_unused = 0
```
Comment (kdreese) [inline src/cards/defense/barricade_3.tres:8]: ??

File: src/cards/card_pool.gd [hunk 14]
```diff
@@ -0,0 +1,25 @@
+extends Resource
+class_name CardPool
```
Comment (thearst3rd) [inline src/cards/card_pool.gd]:
```gdscript
class_name CardPool
extends Resource
```

File: src/cards/card_pool_entry.gd [hunk 15]
```diff
@@ -0,0 +1,7 @@
+extends Resource
+class_name CardPoolEntry
```
Comment (thearst3rd) [inline src/cards/card_pool_entry.gd]: same here

**Review by kdreese — CHANGES_REQUESTED** [review]

File: src/cards/dual_card.gd (hunk 2 above)
Comment (kdreese) [reply src/cards/dual_card.gd:76]: Nope, it's bitwise

**Review by thearst3rd — CHANGES_REQUESTED** [review]
tiny nitpicks

**Review by kdreese — DISMISSED** [review]

**Review by thearst3rd — DISMISSED** [review]

File: src/cards/actions/charge.gd (hunk 1 above)
Comment (thearst3rd) [reply src/cards/actions/charge.gd]: I would also say yes

**Review by kdreese — APPROVED** [review]

**Review by thearst3rd — APPROVED** [review]
